File: 백준/Platinum/1786. 찾기/찾기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.List;

public class Main {
    static int N, M;
    static char[] text, pattern;
    static int[] pi;
    static List<Integer> answer = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        text = br.readLine().toCharArray();
        N = text.length;
        pattern = br.readLine().toCharArray();
        M = pattern.length;

        makePi();
        setAnswer();
        System.out.println(answer.size());
        for (int ans : answer)
            System.out.print(ans + " ");
    }

    static void makePi() {
        pi = new int[M];

        for (int i = 1, j = 0; i < M; i++) {
            while (j > 0 && pattern[i] != pattern[j]) {
                j = pi[j - 1];
            }
            if (pattern[i] == pattern[j]) {
                pi[i] = ++j;
            }
        }
    }

    static void setAnswer() {
        for (int i = 0, j = 0; i < N; i++) {
            while (j > 0 && text[i] != pattern[j]) {
                j = pi[j - 1];
            }

            if (text[i] == pattern[j]) {
                if (j == M - 1) {
                    answer.add(i - j + 1);
                    j = pi[j];
                } else {
                    j++;
                }
            }
        }
    }
}

File: 백준/Platinum/1786. 찾기/찾기.py
def make_pi(pattern):
    M = len(pattern)
    pi = [0] * M
    j = 0
    for i in range(1, M):
        while j > 0 and pattern[i] != pattern[j]:
            j = pi[j-1]
        if pattern[i] == pattern[j]:
            j += 1
            pi[i] = j
    return pi

def kmp(text, pattern):
    N = len(text)
    M = len(pattern)
    pi = make_pi(pattern)
    answer = []
    j = 0
    for i in range(N):
        while j > 0 and text[i] != pattern[j]:
            j = pi[j-1]
        if text[i] == pattern[j]:
            if j == M-1:
                answer.append(i - M + 2) # 자바와 다르게 파이썬은 0이 아닌 1부터 시작하는 인덱스를 사용합니다.
                j = pi[j]
            else:
                j += 1
    return answer

text = input()
pattern = input()
answer = kmp(text, pattern)
print(len(answer))
for ans in answer:
    print(ans, end=' ')